kmod/src/omap.c

```diff
@@ -0,0 +1,1062 @@
+	struct omap_request *req;
+	u64 resp_rid;
+	u64 resp_id;
+	int ret;
+
+	map = kmalloc(sizeof(struct scoutfs_open_ino_map), GFP_NOFS);
+	if (!map) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	rcu_read_lock();
+	req = rhashtable_lookup(&ominf->req_ht, &resp_map->args.req_id, req_ht_params);
+	if (req) {
+		spin_lock(&req->lock);
+		entry = find_rid(&req->rids, rid);
+		if (entry) {
+			bitmap_or((unsigned long *)req->map.bits, (unsigned long *)req->map.bits,
+				  (unsigned long *)resp_map->bits, SCOUTFS_OPEN_INO_MAP_BITS);
+			if (free_rid(&req->rids, entry) == 0)
+				send_response = true;
+		}
+		spin_unlock(&req->lock);
+
+		if (send_response) {
+			resp_rid = req->client_rid;
+			resp_id = req->client_id;
+			memcpy(map, &req->map, sizeof(struct scoutfs_open_ino_map));
+			rhashtable_remove_fast(&ominf->req_ht, &req->ht_head, req_ht_params);
+			call_rcu(&req->rcu, free_req_rcu);
+		}
+	}
+	rcu_read_unlock();
+
+	if (send_response)
+		ret = scoutfs_server_send_omap_response(sb, resp_rid, resp_id, map, 0);
+	else
+		ret = 0;
+	kfree(map);
+out:
+	return ret;
+}
+
+/*
+ * The server is shutting down.  Free all the server state associated
+ * with ongoing request processing.  Clients who still have requests
+ * pending will resend them to the next server.
+ */
+void scoutfs_omap_server_shutdown(struct super_block *sb)
+{
+	DECLARE_OMAP_INFO(sb, ominf);
+	struct rhashtable_iter iter;
+	struct llist_node *requests;
+	struct omap_request *req;
+	struct omap_request *tmp;
+
+	rhashtable_walk_enter(&ominf->req_ht, &iter);
+	rhashtable_walk_start(&iter);
+
+	for (;;) {
+		req = rhashtable_walk_next(&iter);
+		if (req == NULL)
+			break;
+		if (req == ERR_PTR(-EAGAIN))
+			continue;
+
+		if (req->rids.nr_rids != 0) {
+			free_rids(&req->rids);
+			rhashtable_remove_fast(&ominf->req_ht, &req->ht_head, req_ht_params);
+			call_rcu(&req->rcu, free_req_rcu);
+		}
+	}
+
+	rhashtable_walk_stop(&iter);
+	rhashtable_walk_exit(&iter);
+
+	requests = llist_del_all(&ominf->requests);
+	llist_for_each_entry_safe(req, tmp, requests, llnode)
+		kfree(req);
+
+	synchronize_rcu();
+}
+
+/*
+ * Make sure that we have an omap_lock for the cluster lock.  Multiple
+ * users can race to allocate a new olk and once its stored our use of
+ * the cluster lock stops it from being freed.
+ */
+static int get_omap_lock(struct scoutfs_lock *lock, struct scoutfs_omap_lock **olk_ret)
+{
+	struct scoutfs_omap_lock *olk = NULL;
+	int ret = 0;
+
+	spin_lock(&lock->omap_spinlock);
+	olk = lock->omap_lock;
+	spin_unlock(&lock->omap_spinlock);
+
+	if (olk == NULL) {
+		olk = kzalloc(sizeof(struct scoutfs_omap_lock), GFP_NOFS);
+		if (olk) {
+			olk->version = lock->write_version - 1; /* ensure refresh */
+			init_waitqueue_head(&olk->waitq);
+
+			spin_lock(&lock->omap_spinlock);
+			if (lock->omap_lock == NULL) {
+				lock->omap_lock = olk;
+			} else {
+				kfree(olk);
+				olk = lock->omap_lock;
+			}
+			spin_unlock(&lock->omap_spinlock);
+		} else {
+			ret = -ENOMEM;
+		}
+	}
```

Comment:
We're dropping the spinlock here because we don't want to hold it while kzallocing a 184-byte struct? If uncontended, overall slightly slower due to locking twice. If contended, each thread kzallocs but one wins, and the others throw their alloc away? If the lock was only acquired once then other threads would be spinning but then would get in and out without more work. And this fn would be simpler.

It really depends on so much -- relative costs of operations, expected degree of contention, etc, that I'm not clear on, so just wanted to understand the thinking a little more.

Reply:
Yeah, we can't use blocking allocs while holding a spinlock.  The unlikely extra alloc/free if two race to alloc is the simple alternative to trying non-blocking allocs or, say, always allocating before acquiring the lock.  But your confusion here made be rethink the different alloc/get functions which didn't make a huge amount of sense.  I mushed them together into a get_current_lock_data which does everything under the spinlock except obviously dropping when it has to block to allocate, send requests, or wait for request.  Hopefully that makes this all a whole lot clearer.